### DISCUSSION TITLE
Multiapps user object based transfer - Seg fault with UserObjectInterface

### DISCUSSION THREAD

**Discussion by abarun22** [idaholab/moose#19322] [Q&A General] [answered]
Dear all,
I am trying to develop an user object based multi-apps transfer mechanism along the lines of `MultiAppUserObjectTransfer`.  The issue being that i could not get the basic framework to execute for the simplest of cases. There seems to be an seg-fault problem with the inclusion of `UserObjectInterface` which is a direct base class of the new transfer class that i am currently developing. Execution is OK without `UserObjectInterface`, however with the cost of undefined reference to `getUserObjectByName` and therefore i continued inheriting this class and moved on to debug the seg-fault. I tried to check this issue in the debug mode and obtained the following trace back. 
<code><pre>
#1  0x00007ffff7bc881a in std::_Rb_tree<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, libMesh::Parameters::Value*>, std::_Select1st<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, libMesh::Parameters::Value*> >, std::less<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, libMesh::Parameters::Value*> > >::find (this=0x311, __k=...) at /usr/local/depot/gcc-7.3.0/include/c++/7.3.0/bits/stl_tree.h:2536
#2  0x00007ffff7bc725b in std::__cxx1998::map<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, libMesh::Parameters::Value*, std::less<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, libMesh::Parameters::Value*> > >::find (this=0x311, __x=...) at /usr/local/depot/gcc-7.3.0/include/c++/7.3.0/bits/stl_map.h:1189
#3  0x00007ffff7bc54f0 in std::__debug::map<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, libMesh::Parameters::Value*, std::less<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, libMesh::Parameters::Value*> > >::find (this=0x2f9, __x=...) at /usr/local/depot/gcc-7.3.0/include/c++/7.3.0/debug/map.h:574
#4  0x00007ffff6a95306 in libMesh::Parameters::have_parameter<FEProblemBase*> (this=0x2f1, name=...)
    at /home/abalasub/Working/Projects/moose/scripts/../libmesh/installed/include/libmesh/parameters.h:407
#5  0x00007ffff6a95a95 in libMesh::Parameters::get<FEProblemBase*> (this=0x2f1, name=...)
    at /home/abalasub/Working/Projects/moose/scripts/../libmesh/installed/include/libmesh/parameters.h:426
#6  0x00007ffff6a9415f in InputParameters::getCheckedPointerParam<FEProblemBase*> (this=0x2f1, name=..., error_string=...)
    at /home/abalasub/Working/Projects/moose/framework/build/header_symlinks/InputParameters.h:1211
#7  0x00007ffff2454c90 in UserObjectInterface::UserObjectInterface (this=0xf6df18, moose_object=0xf6d800)
    at /home/abalasub/Working/Projects/moose/framework/src/userobject/UserObjectInterface.C:25
#8  0x00007ffff1cd51f3 in MultiAppTransferQuantity_UO::MultiAppTransferQuantity_UO (this=0xf6d800, parameters=...,
    __in_chrg=<optimized out>, __vtt_parm=<optimized out>)
    at /home/abalasub/Working/Projects/moose/framework/src/transfers/MultiAppTransferQuantity_UO.C:33
#9  0x00007ffff1d5782d in __gnu_cxx::new_allocator<MultiAppTransferQuantity_UO>::construct<MultiAppTransferQuantity_UO, InputParameters const&> (this=0x7fffffffbc27, __p=0xf6d800, __args#0=...)
#10 0x00007ffff1d54e84 in std::allocator_traits<std::allocator<MultiAppTransferQuantity_UO> >::construct<MultiAppTransferQuantity_UO, InputParameters const&> (__a=..., __p=0xf6d800, __args#0=...)
    at /usr/local/depot/gcc-7.3.0/include/c++/7.3.0/bits/alloc_traits.h:475
#11 0x00007ffff1d50cff in std::_Sp_counted_ptr_inplace<MultiAppTransferQuantity_UO, std::allocator<MultiAppTransferQuantity_UO>, (__gnu_cxx::_Lock_policy)2>::_Sp_counted_ptr_inplace<InputParameters const&> (this=0xf6d7f0, __a=...)
    at /usr/local/depot/gcc-7.3.0/include/c++/7.3.0/bits/shared_ptr_base.h:526
#12 0x00007ffff1d4befb in std::__shared_count<(__gnu_cxx::_Lock_policy)2>::__shared_count<MultiAppTransferQuantity_UO, std::al---Type <return> to continue, or q <return> to quit---
locator<MultiAppTransferQuantity_UO>, InputParameters const&> (this=0x7fffffffbdb8, __a=...)
    at /usr/local/depot/gcc-7.3.0/include/c++/7.3.0/bits/shared_ptr_base.h:637
#13 0x00007ffff1d46686 in std::__shared_ptr<MultiAppTransferQuantity_UO, (__gnu_cxx::_Lock_policy)2>::__shared_ptr<std::allocator<MultiAppTransferQuantity_UO>, InputParameters const&> (this=0x7fffffffbdb0, __tag=..., __a=...)
    at /usr/local/depot/gcc-7.3.0/include/c++/7.3.0/bits/shared_ptr_base.h:1295
#14 0x00007ffff1d3efd7 in std::shared_ptr<MultiAppTransferQuantity_UO>::shared_ptr<std::allocator<MultiAppTransferQuantity_UO>, InputParameters const&> (this=0x7fffffffbdb0, __tag=..., __a=...)
    at /usr/local/depot/gcc-7.3.0/include/c++/7.3.0/bits/shared_ptr.h:344
#15 0x00007ffff1d33fc7 in std::allocate_shared<MultiAppTransferQuantity_UO, std::allocator<MultiAppTransferQuantity_UO>, InputParameters const&> (__a=..., __args#0=...) at /usr/local/depot/gcc-7.3.0/include/c++/7.3.0/bits/shared_ptr.h:691
#16 0x00007ffff1d2251f in std::make_shared<MultiAppTransferQuantity_UO, InputParameters const&> (__args#0=...)
    at /usr/local/depot/gcc-7.3.0/include/c++/7.3.0/bits/shared_ptr.h:707
#17 0x00007ffff1d0d8d0 in moose::internal::buildObj<MultiAppTransferQuantity_UO> (parameters=...)
    at /home/abalasub/Working/Projects/moose/framework/build/header_symlinks/Registry.h:146
#18 0x00007ffff2bc6187 in Factory::create (this=0x7ffff7ebb8f0, obj_name=..., name=..., parameters=..., tid=0,
    print_deprecated=false) at /home/abalasub/Working/Projects/moose/framework/src/base/Factory.C:131
#19 0x00007ffff1912324 in Factory::create<Transfer> (this=0x7ffff7ebb8f0, obj_name=..., name=..., parameters=..., tid=0)
    at /home/abalasub/Working/Projects/moose/framework/build/header_symlinks/Factory.h:303
#20 0x00007ffff18b901b in FEProblemBase::addTransfer (this=0xd97660, transfer_name=..., name=..., parameters=...)
    at /home/abalasub/Working/Projects/moose/framework/src/problems/FEProblemBase.C:4482
#21 0x00007ffff2340c8f in AddTransferAction::act (this=0xa9cfc0)
    at /home/abalasub/Working/Projects/moose/framework/src/actions/AddTransferAction.C:30
#22 0x00007ffff232f5ad in Action::timedAct (this=0xa9cfc0)
    at /home/abalasub/Working/Projects/moose/framework/src/actions/Action.C:93
#23 0x00007ffff2334023 in ActionWarehouse::executeActionsWithAction (this=0x7ffff7ebb510, task=...)
    at /home/abalasub/Working/Projects/moose/framework/src/actions/ActionWarehouse.C:392
#24 0x00007ffff2333acd in ActionWarehouse::executeAllActions (this=0x7ffff7ebb510)
    at /home/abalasub/Working/Projects/moose/framework/src/actions/ActionWarehouse.C:352
#25 0x00007ffff2b1abd5 in MooseApp::runInputFile (this=0x7ffff7e6c020)
    at /home/abalasub/Working/Projects/moose/framework/src/base/MooseApp.C:1017
#26 0x00007ffff2b1d600 in MooseApp::run (this=0x7ffff7e6c020)
    at /home/abalasub/Working/Projects/moose/framework/src/base/MooseApp.C:1281
#27 0x0000000000412c91 in main (argc=3, argv=0x7fffffffda98) at /home/abalasub/Working/Projects/submodel_repo/src/main.C:33
</code></pre>

A detailed analysis of traceback reveals the problem runs very deep in to the libmesh functionalities and that it is harder to find the exact cause of this issue. I wonder if there is any we could do here to find the right fix for this issue. Attached here are the modified sources and the input files used for this simulation, which might possibly throw some hints as to check where things are going wrong. Any suggestions here are highly appreciated.
Kind regards,
Arun

[08112021.zip](https://github.com/idaholab/moose/files/7499414/08112021.zip)

**Comment by GiudGiud** [comment, marked as answer]
Hello

It seems the issue you are getting is because of this InputParameters::getCheckedPointerParam<FEProblemBase*>
This is used by the UO interface to get access to the FEProblem. i would put some print statements there to understand when this is being run, maybe it s being ran too early and the problem has not been created yet. I doubt it s that though, Transfers should be created fairly late. 

I think you have moved forward on this, based on this other discussion https://github.com/idaholab/moose/discussions/19485

Guillaume
